src/lib_file_deps.ml

```diff
@@ -1,79 +1,51 @@
 open Stdune
-open Dune_file
-open Build_system
 
 module Group = struct
   type t =
     | Cmi
     | Cmx
     | Header
 
+  let all = [Cmi; Cmx; Header]
+
   let to_string = function
     | Cmi -> ".cmi"
     | Cmx -> ".cmx"
     | Header -> ".h"
 
-  let of_cm_kind = function
-    | Cm_kind.Cmx -> Cmx
-    | Cmi -> Cmi
-    | Cmo -> Exn.code_error "Lib_file_deps.Group.of_cm_kind: Cmo" []
-
-  module L = struct
-    let to_string l =
-      List.map l ~f:to_string
-      |> List.sort ~compare:String.compare
-      |> String.concat  ~sep:"-and-"
-
-    let alias t ~dir ~name =
-      sprintf "lib-%s%s-all" (Lib_name.to_string name) (to_string t)
-      |> Alias.make ~dir
-
-    let setup_alias t ~dir ~lib ~files =
-      Build_system.Alias.add_deps
-        (alias t ~dir ~name:(Library.best_name lib))
-        files
-
-    let setup_file_deps_group_alias t ~dir ~lib =
-      setup_alias t ~dir ~lib ~files:(
-        List.map t ~f:(fun t ->
-          Alias.stamp_file (alias [t] ~dir ~name:(Library.best_name lib)))
-        |> Path.Set.of_list
-      )
-  end
-end
-
-let setup_file_deps =
-  let cm_kinds = [Cm_kind.Cmx; Cmi] in
-  let groups = List.map ~f:Group.of_cm_kind cm_kinds in
-  fun ~dir ~lib ~modules ->
-    let add_cms ~cm_kind =
-      List.fold_left ~f:(fun acc m ->
-        match Module.cm_public_file m cm_kind with
-        | None -> acc
-        | Some fn -> Path.Set.add acc fn)
+  let obj_dirs t obj_dir =
+    match t with
+    | Cmi -> [Obj_dir.byte_dir obj_dir; Obj_dir.public_cmi_dir obj_dir]
+    | Cmx -> [Obj_dir.native_dir obj_dir]
+    | Header -> [Obj_dir.dir obj_dir]
+
+  let to_predicate =
+    let preds = List.map all ~f:(fun g ->
+      let ext = to_string g in
+      (* we cannot use globs because of bootstrapping. But we reuse the same
```

Comment:
Uhm, why do we need to do this? Just memoize a tiny bit more?

This creates an obligation to prove that the two predicates are equivalent, which seems hard because globs are hard.

Also, your glob seems to be "*exe" rather than "*.exe" unless I'm missing something obvious? What's up with that?

Reply:
It looks like these are not equivalent, by the way :-)

Reply:
> Uhm, why do we need to do this? Just memoize a tiny bit more?

Yeah, basically. Actually, we could have easily just used globs if it wasn't for the damned bootstrap.

> It looks like these are not equivalent, by the way :-)

Another hint please :)

Reply:
```
  | "*"  { glob (Re (seq [no_slash_no_dot; rep no_slash], Bottom)) lexbuf }
```
🤔

Reply:
Ah I see, there's a bit of a subtlety here. Okay, I'll get rid of it this "optimization".